vowpalwabbit/core/include/vw/core/thread_pool.h

```diff
@@ -0,0 +1,187 @@
+// individual contributors. All rights reserved. Released under a BSD (revised)
+// license as described in the file LICENSE.
+#pragma once
+
+#include "vw/core/memory.h"
+
+#include <atomic>
+#include <condition_variable>
+#include <functional>
+#include <future>
+#include <memory>
+#include <mutex>
+#include <queue>
+#include <thread>
+
+// thread pool based off of the simple pool implementation from ccia:
+// https://github.com/anthonywilliams/ccia_code_samples
+
+namespace VW
+{
+template <typename T>
+class thread_safe_queue
+{
+public:
+  thread_safe_queue() {}
+  void push(T value)
+  {
+    std::lock_guard<std::mutex> lk(_mut);
+    _queue.push(std::move(value));
+  }
+
+  bool try_pop(T& value)
+  {
+    std::lock_guard<std::mutex> lk(_mut);
+    if (_queue.empty()) return false;
+    value = std::move(_queue.front());
+    _queue.pop();
+    return true;
+  }
+
+  bool empty() const
+  {
+    std::lock_guard<std::mutex> lk(_mut);
+    return _queue.empty();
+  }
+
+private:
+  mutable std::mutex _mut;
+  std::queue<T> _queue;
+};
+
+class threads_joiner
+{
+public:
+  explicit threads_joiner(std::vector<std::thread>& threads) : _threads{threads} {}
+  ~threads_joiner()
+  {
+    for (std::thread& thread : _threads)
+    {
+      if (thread.joinable()) thread.join();
+    }
+  }
+
+private:
+  std::vector<std::thread>& _threads;
+};
+
+class thread_pool
+{
+private:
+  class function_wrapper
+  {
+    // function wrapper is a type erased move-only function storer needed in order to temporarily store the submitted
+    // task
+    // in a std::packaged_task which in turn is move-only
+  public:
+    template <typename F>
+    function_wrapper(F&& f) : _impl{VW::make_unique<impl_type<F>>(std::forward<F>(f))}
+    {
+    }
+
+    function_wrapper() = default;
+
+    function_wrapper(function_wrapper&& other) : _impl{std::move(other._impl)} {}
+
+    function_wrapper(const function_wrapper&) = delete;
+
+    function_wrapper(function_wrapper&) = delete;
+
+    function_wrapper& operator=(const function_wrapper&) = delete;
+
+    void operator()() { _impl->call(); }
+
+    function_wrapper& operator=(function_wrapper&& other)
+    {
+      _impl = std::move(other._impl);
+      return *this;
+    }
+
+  private:
+    struct impl_base
+    {
+      virtual void call() = 0;
+      virtual ~impl_base() {}
+    };
+
+    template <typename F>
+    struct impl_type : impl_base
+    {
+      F _f;
+      impl_type(F&& f) : _f{std::forward<F>(f)} {}
+      void call() { _f(); }
+    };
+    std::unique_ptr<impl_base> _impl;
+  };
+
+  void worker()
+  {
+    while (!_done)
+    {
+      function_wrapper task;
+      {
+        std::unique_lock<std::mutex> l(_mut);
+        while (!_task_queue.try_pop(task) && !_done) { _cv.wait(l); }
+      }
+      if (_done) { break; }
+      task();
+    }
+  }
+
+  std::atomic_bool _done;
+  std::condition_variable _cv;
+  mutable std::mutex _mut;

```

Comment:
I dont think this needs to be mutable

Reply:
that's true it isn't used in any read only functions